### ISSUE TITLE
[FSDP2][NCCL] Large slowdown for non-16-byte-aligned per-rank collective payloads

### ISSUE THREAD

**Issue by 0x777a6c** [pytorch/pytorch#195450]
## Summary

Using PyTorch's NCCL tensor collectives on a flat buffer with the same per-rank layout as FSDP2's default non-symmetric-memory communication path, a per-rank payload whose byte count is not a multiple of 16 causes a substantial latency increase. The effect is reproducible on both a single node and a two-node RoCE setup.

The attached reproducer directly tests `all_gather_into_tensor` and `reduce_scatter_tensor`; it does not instantiate an FSDP2 model. These APIs exercise the same single-tensor NCCL collective path used by FSDP2.

This report concerns the regular, non-symmetric-memory NCCL path. It does not change tensor values, parameter shapes, or model architecture.

## Environment

| Component | Configuration |
| --- | --- |
| GPU | NVIDIA RTX PRO 6000 Blackwell |
| NIC | NVIDIA ConnectX-7 |
| Network | RoCE |
| Topology | 2 nodes x 8 GPUs; 1 node x 8 GPUs |
| PyTorch | 2.8.0+cu128 |
| CUDA | 12.8 |
| NCCL | 2.27.3 |
| NVIDIA driver | 595.71.05 |
| Backend | NCCL |
| Dtype | `bfloat16` |

For the two-node runs, the following NCCL variables were set on both nodes:

```bash
export NCCL_IB_QPS_PER_CONNECTION=4
export NCCL_MIN_NCHANNELS=8
```

The benchmark allocates one maximum-size input buffer and one maximum-size output buffer per rank, then reuses the same GPU pointers for every count. Only the tensor views and counts change, so allocation and pointer changes are not part of the comparison.

## Results

`collective_mean_ms` is the mean over 10 measured iterations. For each iteration, the reported value is the maximum per-rank CUDA-event duration.

| Count (elements) | Per-rank payload (B) | Payload mod 16 | 2-node `all_gather` (ms) | 2-node `reduce_scatter` (ms) | 1-node `all_gather` (ms) | 1-node `reduce_scatter` (ms) |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 62,662,200 | 125,324,400 | 0 | 46.58 | 46.84 | 22.75 | 22.07 |
| 62,662,192 | 125,324,384 | 0 | 47.11 | 47.15 | 22.48 | 22.10 |
| 62,662,197 | 125,324,394 | 10 | 140.30 | 72.44 | 63.22 | 34.09 |

`Payload mod 16` is computed from the per-rank chunk stride, not from the base pointer or total buffer size. The total buffer can still be divisible by 16 when the per-rank stride is not.

The unaligned case is bracketed by two aligned cases whose per-rank payload sizes differ by only a few bytes, helping isolate alignment from message-size threshold effects.

## Reproduction

**Attachment:** [collective_alignment_measure.py](https://github.com/user-attachments/files/31647079/collective_alignment_measure.py)

Two nodes:

```bash
torchrun \
  --nnodes=2 \
  --nproc_per_node=8 \
  --node_rank=<0-or-1> \
  --master_addr=<MASTER_ADDR> \
  --master_port=29771 \
  ./collective_alignment_measure.py \
  --op all_gather reduce_scatter \
  --counts 62662200 62662192 62662197 \
  --warmups 10 \
  --iters 10
```

One node:

```bash
torchrun \
  --standalone \
  --nproc_per_node=8 \
  ./collective_alignment_measure.py \
  --op all_gather reduce_scatter \
  --counts 62662200 62662192 62662197 \
  --warmups 3 \
  --iters 10
```

For per-iteration diagnostics, append `--dump-samples`; this adds the maximum time across ranks for each measured iteration to the corresponding result line.

## FSDP2 impact and proposed fix

FSDP2 packs each parameter group into a flat communication buffer. The NCCL per-rank stride is the total logical packed shard size. When that stride is not 16-byte aligned, the starting address of each rank's chunk cycles through different 16-byte remainders.

The proposed fix rounds only the communication staging stride up to a 16-byte boundary while keeping logical tensor sizes unchanged:

- **All-gather:** add the padding in the existing copy-in operation, run NCCL with the padded per-rank stride, and use the existing copy-out operation to copy only the logical elements.
- **Reduce-scatter:** extend the existing rank-major gradient pack so that each rank's input segment has a 16-byte-aligned stride. After NCCL, expose only the logical output range and ignore the trailing padding.

The change reuses the existing copy-in, copy-out, and gradient-pack operations; it does not introduce another full-buffer copy or alter parameter/gradient shapes. The padding is confined to communication staging buffers and is at most 15 bytes per rank segment. The proposed change targets the default non-symmetric-memory NCCL path; the symmetric-memory path can retain its existing behavior.

## Related issue

This is related to [PyTorch #187239](https://github.com/pytorch/pytorch/issues/187239). That issue targets a per-rank 16-byte-aligned `all_gather` layout for symmetric-memory constraints and copy-out elimination. This report is independent and applies the same alignment principle to the default non-symmetric-memory NCCL `all_gather` and `reduce_scatter` paths for collective performance.

Would the maintainers accept a PR implementing this for the default NCCL path? I can provide focused correctness tests and benchmarks.

### Versions

PyTorch version: 2.8.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: Could not collect
CMake version: version 4.1.3
Libc version: glibc-2.35

Python version: 3.12.8 | packaged by conda-forge | (main, Dec  5 2024, 14:24:40) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.14.0-284.11.1.el9_2.x86_64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA RTX PRO 6000 Blackwell Server Edition
GPU 1: NVIDIA RTX PRO 6000 Blackwell Server Edition
GPU 2: NVIDIA RTX PRO 6000 Blackwell Server Edition
GPU 3: NVIDIA RTX PRO 6000 Blackwell Server Edition
GPU 4: NVIDIA RTX PRO 6000 Blackwell Server Edition
GPU 5: NVIDIA RTX PRO 6000 Blackwell Server Edition
GPU 6: NVIDIA RTX PRO 6000 Blackwell Server Edition
GPU 7: NVIDIA RTX PRO 6000 Blackwell Server Edition

Nvidia driver version: 595.71.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: {'PYTORCH_CUDA_ALLOC_CONF': 'expandable_segments:True'}

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   52 bits physical, 57 bits virtual
Byte Order:                      Little Endian
CPU(s):                          180
On-line CPU(s) list:             0-179
Vendor ID:                       GenuineIntel
BIOS Vendor ID:                  QEMU
Model name:                      INTEL(R) XEON(R) PLATINUM 8558P
BIOS Model name:                 pc-q35-6.1
CPU family:                      6
Model:                           207
Thread(s) per core:              2
Core(s) per socket:              45
Socket(s):                       2
Stepping:                        2
BogoMIPS:                        5400.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 wbnoinvd arat avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b fsrm md_clear serialize tsxldtrk amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       4.2 MiB (90 instances)
L1i cache:                       2.8 MiB (90 instances)
L2 cache:                        180 MiB (90 instances)
L3 cache:                        520 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-89
NUMA node1 CPU(s):               90-179
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Vulnerable
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.5.1
[pip3] numpydantic==1.8.0
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.3
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] torch==2.8.0
[pip3] torchaudio==2.8.0+cu128
[pip3] torchcodec==0.6.0
[pip3] torchdata==0.11.0
[pip3] torchvision==0.23.0
[pip3] triton==3.4.0
[conda] numpy                     2.5.1                    pypi_0    pypi
[conda] numpydantic               1.8.0                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.8.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.8.90                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.8.93                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.8.90                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.10.2.21                pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.3.83                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.9.90                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.3.90                pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.8.93                pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.7.1                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.27.3                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.8.93                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.8.90                  pypi_0    pypi
[conda] torch                     2.8.0                    pypi_0    pypi
[conda] torchaudio                2.8.0+cu128              pypi_0    pypi
[conda] torchcodec                0.6.0                    pypi_0    pypi
[conda] torchdata                 0.11.0                   pypi_0    pypi
[conda] torchvision               0.23.0                   pypi_0    pypi
[conda] triton                    3.4.0                    pypi_0    pypi

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @kapilsh

**Comment by weifengpy** [comment]
thanks for the detailed report!

**Comment by weifengpy** [comment]
we are doing per-fqn mixed precision https://github.com/pytorch/pytorch/issues/156784. might worth triage this together. assigned to @anshul-si

**Comment by ngimel** [comment]
You will find big slowdowns on unaligned inputs all across pytorch operations, and that's expected. All those operations don't exist in vacuum, they exist to feed tensorcores, and tensorcores only accept aligned addresses

**Comment by 0x777a6c** [comment]
> we are doing per-fqn mixed precision [#156784](https://github.com/pytorch/pytorch/issues/156784). might worth triage this together. assigned to [@anshul-si](https://github.com/anshul-si)

Thanks! Happy to coordinate with @anshul-si

**Comment by 0x777a6c** [comment]
> You will find big slowdowns on unaligned inputs all across pytorch operations, and that's expected. All those operations don't exist in vacuum, they exist to feed tensorcores, and tensorcores only accept aligned addresses

In this case, the base address is 16-byte aligned. It's the unaligned per-rank offsets that cause a substantial slowdown in the NCCL collective.

**Comment by weifengpy** [comment]
my assumption here is fsdp2 have control over how params are copied into a new address, and whether address are 16-byte aligned.
for regular nn.Module params, pytorch don't have control over param shapes (thus address)

**Comment by 0x777a6c** [comment]
> my assumption here is fsdp2 have control over how params are copied into a new address, and whether address are 16-byte aligned. for regular nn.Module params, pytorch don't have control over param shapes (thus address)

That makes sense. Since FSDP2 controls the staging layout, may I prepare a focused PR to align the per-rank all-gather and reduce-scatter staging strides?

**Comment by weifengpy** [comment]
> That makes sense. Since FSDP2 controls the staging layout, may I prepare a focused PR to align the per-rank all-gather and reduce-scatter staging strides?

@0x777a6c can you build on top of per-fqn mixed precision? https://github.com/pytorch/pytorch/pull/196626/ 
also cc @anshul-si : do we still have the issue after per-fqn mixed precision ?
